File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        Node *current=head;
        map<Node*,Node*>mp;
        Node *temp;
        while(current!=NULL)
        {   temp=new Node(current->val);
            mp[current]=temp;
            current=current->next;
        }
        current=head;
        while(current!=NULL)
        {
            mp[current]->next=mp[current->next];
            mp[current]->random=mp[current->random];
            current=current->next;
        }
        return mp[head];
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    h := make(map[*Node]*Node)
    cur := head
    for cur != nil{
        h[cur] = &Node{
            Val: cur.Val,
        }
        cur = cur.Next
    }
    cur = head
    for cur != nil {
        h[cur].Next = h[cur.Next]
        h[cur].Random = h[cur.Random]
        cur = cur.Next
    }

    return h[head]
}